vllm/distributed/parallel_state.py

```diff
@@ -162,12 +176,33 @@ def initialize_model_parallel(
             _TP_DEVICE_GROUP = group
             _TP_CPU_GROUP = cpu_group
 
+    # A small all_reduce for warmup.
+    # this is needed because device communicators might be created lazily
+    # (e.g. NCCL). This will ensure that the communicator is initialized
+    # before any communication happens, so that this group can be used for
+    # graph capture immediately.
+    data = torch.zeros(1)
+    if torch.cuda.is_available():
+        data = data.to(device=f"cuda:{_LOCAL_RANK}")
+    torch.distributed.all_reduce(data, group=_TP_DEVICE_GROUP)
+    if torch.cuda.is_available():
+        torch.cuda.synchronize()
+
     from vllm.distributed.device_communicators.pynccl import PyNcclCommunicator
     _TP_PYNCCL_COMMUNICATOR = PyNcclCommunicator(
         group=_TP_CPU_GROUP,
         device=_LOCAL_RANK,
     )
 
+    # Initialize a custom fast all-reduce implementation.
+    if _ENABLE_CUSTOM_ALL_REDUCE:
+        from vllm.distributed.device_communicators.custom_all_reduce import (
```

Comment:
Why do we need lazy import here?

Reply:
The circular import:

`vllm/distributed/__init__.py` imports `parallel_state` and `communication_op`. If `parallel_state` imports from `vllm.distributed.device_communicators.custom_all_reduce` in the top level, then this is a circular import because `custom_all_reduce` imports `get_tensor_model_parallel_cpu_group` from `parallel_state`.

Therefore, either `parallel_state` or `custom_all_reduce` has to use lazy import to break the circular import.

I use lazy import in `parallel_state`, to be consistent with how we import `pynccl`.